source/dub/logging.d

```diff
@@ -0,0 +1,384 @@
+{
+	log(LogLevel.info, true, "", Color.init, fmt, args);
+}
+
+/**
+	Shorthand function to log a message with warning level, with custom tag string.
+	The tag color is fixed to yellow.
+
+	Params:
+		tag = The string the tag at the beginning of the line should contain
+		level = The log level for the logged message
+		fmt = See http://dlang.org/phobos/std_format.html#format-string
+*/
+void logWarnTag(T...)(string tag, string fmt, lazy T args) nothrow
+{
+	log(LogLevel.warn, false, tag, Color.yellow, fmt, args);
+}
+
+/**
+	Shorthand function to log a message with warning level, using the default
+	tag "Warning". The tag color is also fixed to yellow.
+
+	Params:
+		level = The log level for the logged message
+		fmt = See http://dlang.org/phobos/std_format.html#format-string
+*/
+void logWarn(T...)(string fmt, lazy T args) nothrow
+{
+	log(LogLevel.warn, false, "Warning", Color.yellow, fmt, args);
+}
+
+/**
+	Shorthand function to log a message with error level, with custom tag string.
+	The tag color is fixed to red.
+
+	Params:
+		tag = The string the tag at the beginning of the line should contain
+		level = The log level for the logged message
+		fmt = See http://dlang.org/phobos/std_format.html#format-string
+*/
+void logErrorTag(T...)(string tag, string fmt, lazy T args) nothrow
+{
+	log(LogLevel.error, false, tag, Color.red, fmt, args);
+}
+
+/**
+	Shorthand function to log a message with error level, using the default
+	tag "Error". The tag color is also fixed to red.
+
+	Params:
+		level = The log level for the logged message
+		fmt = See http://dlang.org/phobos/std_format.html#format-string
+*/
+void logError(T...)(string fmt, lazy T args) nothrow
+{
+	log(LogLevel.error, false, "Error", Color.red, fmt, args);
+}
+
+/**
+	Log a message with the specified log level and with the specified tag string
+	and color. If the log level is debug or diagnostic, the tag is not printed
+	thus the tag string and tag color will be ignored. If the log level is error
+	or warning, the tag will be in bold text. Also the tag can be disabled (for
+	any log level) by passing true as the second argument.
+
+	Params:
+		level = The log level for the logged message
+		disableTag = Setting this to true disables the tag, no matter what
+		tag = The string the tag at the beginning of the line should contain
+		tagColor = The color the tag string should have
+		fmt = See http://dlang.org/phobos/std_format.html#format-string
+*/
+void log(T...)(
+	LogLevel level,
+	bool disableTag,
+	string tag,
+	Color tagColor,
+	string fmt,
+	lazy T args
+) nothrow
+{
+	if (level < _minLevel)
+		return;
+
+	auto hasTag = true;
+	if (level <= LogLevel.diagnostic)
+		hasTag = false;
+	if (disableTag)
+		hasTag = false;
+
+	auto boldTag = false;
+	if (level >= LogLevel.warn)
+		boldTag = true;
+
+	try
+	{
+		string result = format(fmt, args);
+
+		if (hasTag)
+			result = tag.rightJustify(TAG_WIDTH, ' ').color(tagColor, boldTag ? Mode.bold : Mode.init) ~ " " ~ result;
```

Comment:
Looks like at least two allocations could be avoided here:
- `formattedWrite` instead of `format`
- direct writes instead of `~`
(`color` + `rightJustify` could be concatenated ranges too).